### PR TITLE
fix(rollup): fix l1DataFee tests

### PR DESCRIPTION
**Pull request by noel2004**
This PR has fixed the unittest `TestCalculateEncodedL1DataFee` for the new routine of l1 fee calculation.

It also add a new test, which should help developer understanding and confirm so details inside l1 fee calculation, like the encoding of tx and reading of L1 gas oracle. Two examples with traces obtained from the a running node has been provided.

### CODE REVIEW

File: rollup/fees/rollup_fee_test.go [hunk 1]
```diff
@@ -14,7 +20,119 @@ func TestCalculateEncodedL1DataFee(t *testing.T) {
...
+	TxDataSample      string
+	EIP1559BaseFee    *big.Int
+	L1basefee         *big.Int
+	L1feeOverHead     *big.Int
+	L1feeScalar       *big.Int
+	EncodedExpected   int
+	L1DataFeeExpected *big.Int
+}
+
+func testCalculateL1DataSize(t *testing.T, t_case *l1DataTestCase) {
```
Comment (0xmountaintop) [inline rollup/fees/rollup_fee_test.go]:
```suggestion
func testEstimateL1DataFeeForTransactionData(t *testing.T, t_case *l1DataTestCase) {
```

File: rollup/fees/rollup_fee_test.go [hunk 2]
```diff
@@ -14,7 +20,119 @@ func TestCalculateEncodedL1DataFee(t *testing.T) {
...
+
+	l1DataFee := calculateEncodedL1DataFee(raw, t_case.L1feeOverHead, t_case.L1basefee, t_case.L1feeScalar)
+	if t_case.L1DataFeeExpected != nil {
+		assert.Equal(t, t_case.L1DataFeeExpected, l1DataFee)
+	} else {
+		t.Log("calculated l1data fee:", l1DataFee)
+	}
+}
+
+func TestCalculateL1DataSize(t *testing.T) {
```
Comment (0xmountaintop) [inline rollup/fees/rollup_fee_test.go]:
```suggestion
func TestEstimateL1DataFeeForTransactionData(t *testing.T) {
```

File: rollup/fees/rollup_fee_test.go [hunk 3]
```diff
@@ -14,7 +20,119 @@ func TestCalculateEncodedL1DataFee(t *testing.T) {
...
+		"from": "0x1c5a77d9fa7ef466951b2f01f724bca3a5820b63",
+		"to": "0x58a2239aa5412f78d8f675c4d8ad5102a3fa5837",
+		"chainId": "0xcf55",
+		"value": "0x0",
+		"data": "0xb0f2b72a000000000000000000000000000000000000000000000000000000000000000a",
+		"isCreate": false
+	}
+`
+
+func reverseDataToMsg(txdata *types.TransactionData) types.Message {
```
Comment (0xmountaintop) [inline rollup/fees/rollup_fee_test.go]:
```suggestion
func transactionDataToMessage(txdata *types.TransactionData) types.Message {
```

File: rollup/fees/rollup_fee_test.go [hunk 4]
```diff
@@ -14,7 +20,119 @@ func TestCalculateEncodedL1DataFee(t *testing.T) {
...
+	databytes, err := hexutil.Decode(txdata.Data)
+	if err != nil {
+		panic(err)
+	}
+	return types.NewMessage(txdata.From, txdata.To, txdata.Nonce, (*big.Int)(txdata.Value), txdata.Gas,
+		(*big.Int)(txdata.GasPrice), (*big.Int)(txdata.GasPrice), (*big.Int)(txdata.GasPrice), databytes, nil, false)
+}
+
+type l1DataTestCase struct {
```
Comment (0xmountaintop) [inline rollup/fees/rollup_fee_test.go]:
```suggestion
type l1DataFeeTestCase struct {
```

File: rollup/fees/rollup_fee_test.go [hunk 5]
```diff
@@ -14,7 +20,119 @@ func TestCalculateEncodedL1DataFee(t *testing.T) {
...
+	txdata := new(types.TransactionData)
+	assert.NoError(t, json.Unmarshal([]byte(t_case.TxDataSample), txdata), "parse json fail")
+
+	// we have decomposed EstimateL1DataFeeForMessage here so
+	// to catch more detail inside the process
+	var msg types.Message
+
+	if t_case.EIP1559BaseFee != nil {
+		//TODO: EIP1559 test
+		panic("no implement")
```
Comment (0xmountaintop) [inline rollup/fees/rollup_fee_test.go]:
```suggestion
		panic("not implemented")
```

File: rollup/fees/rollup_fee_test.go [hunk 6]
```diff
@@ -14,7 +21,119 @@ func TestCalculateEncodedL1DataFee(t *testing.T) {
...
+	chainID := (*big.Int)(txdata.ChainId)
+	signer := types.NewLondonSigner(chainID)
+	unsigned := asUnsignedTx(msg, t_case.EIP1559BaseFee, chainID)
+
+	tx, err := unsigned.WithSignature(signer, append(bytes.Repeat([]byte{0xff}, crypto.SignatureLength-1), 0x01))
+	assert.NoError(t, err, "build dummy tx fail")
+	raw, err := rlpEncode(tx)
+	assert.NoError(t, err, "rlp fail")
+
+	if t_case.EncodedExpected != 0 {
```
Comment (colinlyguo) [inline rollup/fees/rollup_fee_test.go:116]: We don't need this check, right? No `0` in test cases. So does `L1DataFeeExpected`.

**Review by colinlyguo — DISMISSED** [review]

File: rollup/fees/rollup_fee_test.go (hunk 6 above)
Comment (noel2004) [reply rollup/fees/rollup_fee_test.go:116]: This act as a "skip" option so you can put an test example without specified a `EncodedExpected` but still pass.

**Comment by 0xmountaintop** [comment]
fixed in https://github.com/scroll-tech/go-ethereum/pull/755

**Pull request closed without merging**